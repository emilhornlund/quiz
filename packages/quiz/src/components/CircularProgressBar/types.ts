export enum CircularProgressBarKind {
  Default,
  Correct,
  Secondary,
}

export enum CircularProgressBarSize {
  Small,
  Medium,
  Large,
}
