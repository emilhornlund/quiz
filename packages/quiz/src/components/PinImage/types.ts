import { QuestionPinTolerance } from '@quiz/common'

export type PinImagePosition = {
  x: number
  y: number
}

export enum PinColor {
  Green,
  Red,
  Blue,
}

export type PinImageValue = {
  x: number
  y: number
  tolerance?: QuestionPinTolerance
  color?: PinColor
}
