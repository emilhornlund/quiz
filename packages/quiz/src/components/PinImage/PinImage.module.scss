@use '../../styles/colors';
@use '../../styles/helpers';
@use '../../styles/variables';

.pinImage {
  width: 100%;
  height: 100%;
  min-width: 0;
  min-height: 0;
}

/* IMPORTANT: this overlay sits INSIDE the ResponsiveImage fitted box.
   It must allow pointer events to receive drags. */
.overlay {
  position: absolute;
  inset: 0;
  width: 100%;
  height: 100%;
  pointer-events: auto;
  touch-action: none;
  z-index: variables.$pin-image-z-index;

  &:active .pin {
    cursor: grabbing;
  }

  /* The pin is centered via transform so size changes don't affect math */
  .pin {
    position: absolute;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%); /* center by its own size */
    border-radius: 50%;
    cursor: grab;
    user-select: none;
    /* don't clip tolerance */
    overflow: visible;

    /* Large translucent circle centered on the pin, *behind* the dot */
    .tolerance {
      position: absolute;
      left: 50%;
      top: 50%;
      transform: translate(-50%, -50%); /* center this circle */
      border-radius: 50%;
      background: colors.$green-2-25;
      outline: 2px dashed colors.$green-2;
      z-index: 0;
      pointer-events: none; /* don't block dragging */
    }

    /* Visible pin head on top */
    .dot {
      display: flex;
      justify-content: center;
      align-items: center;
      position: relative;
      width: 100%;
      height: 100%;
      border-radius: 50%;
      transform: translate(0%, -50%);
      z-index: 1;

      svg {
        width: auto;
        max-height: 100%;
        height: 100%;
      }
    }
  }

  .children {
    position: relative;
    z-index: 1;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    width: 100%;
    height: 100%;
  }
}
