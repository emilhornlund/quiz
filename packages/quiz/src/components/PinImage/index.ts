export type { PinImageProps } from './PinImage.tsx'
export type { PinImagePosition, PinImageValue } from './types.ts'
export { PinColor } from './types.ts'
export { default } from './PinImage.tsx'
