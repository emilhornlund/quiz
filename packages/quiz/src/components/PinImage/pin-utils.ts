import colors from '../../styles/colors.module.scss'

import { PinColor } from './types.ts'

const roundTo = (v: number, decimals: number) =>
  Math.round((v + Number.EPSILON) * Math.pow(10, decimals)) /
  Math.pow(10, decimals)

export const clamp01 = (v: number) => roundTo(Math.max(0, Math.min(1, v)), 5)

export function getPinColorColor(color: PinColor): string {
  switch (color) {
    case PinColor.Red:
      return colors.red2
    case PinColor.Green:
      return colors.green2
    case PinColor.Blue:
    default:
      return colors.blue2
  }
}
