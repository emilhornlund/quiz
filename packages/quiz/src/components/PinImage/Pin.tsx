import { faLocationDot } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import React, { FC } from 'react'

import { getPinColorColor } from './pin-utils.ts'
import styles from './PinImage.module.scss'
import { PinColor } from './types.ts'

interface PinProps {
  width: number
  height: number
  x: number
  y: number
  toleranceDiameterPx?: number
  color?: PinColor
  disabled?: boolean
}

const Pin: FC<PinProps> = ({
  width,
  height,
  x,
  y,
  toleranceDiameterPx = 0,
  color = PinColor.Blue,
  disabled,
}) => (
  <div
    className={styles.pin}
    style={{
      width,
      height,
      left: `${x * 100}%`,
      top: `${y * 100}%`,
      ...(disabled ? { cursor: 'default' } : {}),
    }}>
    {toleranceDiameterPx > 0 && (
      <div
        className={styles.tolerance}
        style={{
          width: `${toleranceDiameterPx}px`,
          height: `${toleranceDiameterPx}px`,
        }}
      />
    )}
    <div className={styles.dot}>
      <FontAwesomeIcon icon={faLocationDot} color={getPinColorColor(color)} />
    </div>
  </div>
)

export default Pin
