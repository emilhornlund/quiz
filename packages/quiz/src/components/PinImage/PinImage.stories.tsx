import { QuestionPinTolerance } from '@quiz/common'
import type { Meta, StoryObj } from '@storybook/react'
import React, { FC, useState } from 'react'

import PinImage, { PinImageProps } from './PinImage.tsx'
import { PinColor } from './types.ts'

const PinImageStoryComponent: FC<PinImageProps> = (props) => {
  const [value, setValue] = useState<{ x: number; y: number } | undefined>(
    props.value,
  )

  return (
    <div style={{ height: '100vh', display: 'flex', minHeight: 0 }}>
      <div style={{ flex: 1, minHeight: 0 }}>
        <PinImage {...props} value={value} onChange={setValue} />
      </div>
    </div>
  )
}

const meta = {
  component: PinImage,
  parameters: { layout: 'fullscreen' },
  render: (props) => <PinImageStoryComponent {...props} />,
} satisfies Meta<typeof PinImage>

export default meta
type Story = StoryObj<typeof meta>

export const Default = {
  args: {
    imageURL:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/f/f7/Europe_laea_location_map.svg/1401px-Europe_laea_location_map.svg.png',
    value: { x: 0.45838, y: 0.35438, tolerance: QuestionPinTolerance.Medium },
  },
} satisfies Story

export const NoPin = {
  name: 'No Pin',
  args: {
    imageURL:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/f/f7/Europe_laea_location_map.svg/1401px-Europe_laea_location_map.svg.png',
  },
} satisfies Story

export const NoTolerance = {
  name: 'No Tolerance',
  args: {
    imageURL:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/f/f7/Europe_laea_location_map.svg/1401px-Europe_laea_location_map.svg.png',
    value: { x: 0.45838, y: 0.35438 },
  },
} satisfies Story

export const Multiple = {
  name: 'Multiple Pins',
  args: {
    imageURL:
      'https://upload.wikimedia.org/wikipedia/commons/thumb/f/f7/Europe_laea_location_map.svg/1401px-Europe_laea_location_map.svg.png',
    value: {
      x: 0.45838,
      y: 0.35438,
      tolerance: QuestionPinTolerance.Medium,
      color: PinColor.Blue,
    },
    values: [
      { x: 0.5, y: 0.45, color: PinColor.Green },
      { x: 0.3, y: 0.25, color: PinColor.Red },
      { x: 0.85, y: 0.75, color: PinColor.Red },
    ],
    disabled: true,
  },
} satisfies Story
