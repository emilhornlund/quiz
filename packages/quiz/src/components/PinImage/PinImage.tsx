import {
  QUESTION_PIN_TOLERANCE_RADIUS,
  QuestionPinTolerance,
} from '@quiz/common'
import React, {
  FC,
  PointerEvent,
  useCallback,
  useEffect,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
} from 'react'

import { DeviceType, useDeviceSizeType } from '../../utils/use-device-size.tsx'
import ResponsiveImage, { ResponsiveImageProps } from '../ResponsiveImage'

import { clamp01 } from './pin-utils.ts'
import Pin from './Pin.tsx'
import styles from './PinImage.module.scss'
import { PinImagePosition, PinImageValue } from './types.ts'

export type PinImageProps = {
  value?: PinImageValue
  values?: PinImageValue[]
  disabled?: boolean
  onChange?: (position: PinImagePosition) => void
  onValid?: (valid: boolean) => void
} & Pick<ResponsiveImageProps, 'imageURL' | 'alt'>

const PinImage: FC<PinImageProps> = ({
  value,
  values,
  disabled = false,
  imageURL,
  alt,
  onChange,
  onValid,
}) => {
  const deviceType = useDeviceSizeType()

  const pinSizePx = useMemo(() => {
    switch (deviceType) {
      case DeviceType.Mobile:
        return 30
      case DeviceType.Tablet:
        return 35
      default:
        return 40
    }
  }, [deviceType])

  const [interactivePinPosition, setInteractivePinPosition] = useState<
    PinImagePosition | undefined
  >(
    value
      ? {
          x: clamp01(value.x),
          y: clamp01(value.y),
        }
      : undefined,
  )

  const [positions, setPositions] = useState<PinImageValue[]>()
  useEffect(() => {
    const positions =
      values?.map((v) => ({
        ...v,
        x: clamp01(v.x),
        y: clamp01(v.y),
      })) || []
    setPositions(positions)
  }, [values])

  const roRef = useRef<ResizeObserver | null>(null)

  const overlayRef = useRef<HTMLDivElement>(null)
  const [overlaySize, setOverlaySize] = useState<{ w: number; h: number }>({
    w: 0,
    h: 0,
  })

  const setOverlayNode = useCallback((node: HTMLDivElement | null) => {
    roRef.current?.disconnect()
    roRef.current = null
    overlayRef.current = node
    if (!node) return

    const setIfChanged = (w: number, h: number) =>
      setOverlaySize((prev) => (prev.w === w && prev.h === h ? prev : { w, h }))

    const measureNow = () => {
      const r = node.getBoundingClientRect()
      const w = Math.round(r.width)
      const h = Math.round(r.height)
      if (w > 0 && h > 0) setIfChanged(w, h)
    }

    const raf = requestAnimationFrame(measureNow)

    if (typeof ResizeObserver !== 'undefined') {
      roRef.current = new ResizeObserver((entries) => {
        const cr = entries[0]?.contentRect
        if (!cr) return
        const w = Math.round(cr.width)
        const h = Math.round(cr.height)
        if (w > 0 && h > 0) setIfChanged(w, h)
      })
      roRef.current.observe(node)
    }

    const onWinResize = () => measureNow()
    window.addEventListener('resize', onWinResize)

    return () => {
      cancelAnimationFrame(raf)
      roRef.current?.disconnect()
      roRef.current = null
      window.removeEventListener('resize', onWinResize)
    }
  }, [])

  useLayoutEffect(() => {
    const el = overlayRef.current
    if (!el) return

    const setIfChanged = (w: number, h: number) =>
      setOverlaySize((prev) => (prev.w === w && prev.h === h ? prev : { w, h }))

    const measure = () => {
      const rect = el.getBoundingClientRect()
      const w = Math.round(rect.width)
      const h = Math.round(rect.height)
      if (w > 0 && h > 0) setIfChanged(w, h)
    }

    const rafId = requestAnimationFrame(measure)

    const ro =
      typeof ResizeObserver !== 'undefined'
        ? new ResizeObserver((entries) => {
            const cr = entries[0]?.contentRect
            if (!cr) return
            const w = Math.round(cr.width)
            const h = Math.round(cr.height)
            if (w > 0 && h > 0) setIfChanged(w, h)
          })
        : null

    ro?.observe(el)

    const onWinResize = () => measure()
    window.addEventListener('resize', onWinResize)

    return () => {
      cancelAnimationFrame(rafId)
      ro?.disconnect()
      window.removeEventListener('resize', onWinResize)
    }
  }, [imageURL])

  const dragRef = useRef<{
    startX: number
    startY: number
    startPx: { x: number; y: number }
  } | null>(null)

  const onPointerDown = (e: PointerEvent<HTMLDivElement>) => {
    const overlay = overlayRef.current
    if (disabled || !interactivePinPosition || !overlay) return

    const rect = overlay.getBoundingClientRect()
    dragRef.current = {
      startX: e.clientX,
      startY: e.clientY,
      startPx: {
        x: interactivePinPosition.x * rect.width,
        y: interactivePinPosition.y * rect.height,
      },
    }
    ;(e.currentTarget as Element).setPointerCapture?.(e.pointerId)
    e.preventDefault()
  }

  const onPointerMove = (e: PointerEvent<HTMLDivElement>) => {
    if (disabled || !interactivePinPosition || !dragRef.current) return
    const overlay = overlayRef.current
    if (!overlay) return

    const rect = overlay.getBoundingClientRect()
    const dx = e.clientX - dragRef.current.startX
    const dy = e.clientY - dragRef.current.startY

    const newPxX = dragRef.current.startPx.x + dx
    const newPxY = dragRef.current.startPx.y + dy

    const newX = clamp01(newPxX / rect.width)
    const newY = clamp01(newPxY / rect.height)

    setInteractivePinPosition((prevState) => ({
      ...(prevState || {}),
      x: newX,
      y: newY,
    }))
    if (!value || value.x !== newX || value.y !== newY) {
      onChange?.({ x: newX, y: newY })
      onValid?.(true)
    }
  }

  const onPointerUp = (e: PointerEvent<HTMLDivElement>) => {
    if (disabled || !interactivePinPosition) return
    dragRef.current = null
    ;(e.currentTarget as Element).releasePointerCapture?.(e.pointerId)
  }

  const getToleranceDiameterPx = useCallback(
    (tolerance?: QuestionPinTolerance) => {
      if (!overlaySize.w || !overlaySize.h || !tolerance) return 0
      const radius =
        QUESTION_PIN_TOLERANCE_RADIUS[tolerance] *
        Math.min(overlaySize.w, overlaySize.h)
      return Math.max(0, radius * 2)
    },
    [overlaySize],
  )

  return (
    <ResponsiveImage imageURL={imageURL} alt={alt}>
      <div
        ref={setOverlayNode}
        className={styles.overlay}
        onPointerDown={onPointerDown}
        onPointerMove={onPointerMove}
        onPointerUp={onPointerUp}>
        {interactivePinPosition && (
          <Pin
            width={pinSizePx}
            height={pinSizePx}
            x={interactivePinPosition.x}
            y={interactivePinPosition.y}
            toleranceDiameterPx={getToleranceDiameterPx(value?.tolerance)}
            color={value?.color}
            disabled={disabled}
          />
        )}

        {positions?.map((p, i) => (
          <Pin
            key={`${p.x}_${p.y}_${i}`}
            width={pinSizePx}
            height={pinSizePx}
            x={p.x}
            y={p.y}
            toleranceDiameterPx={getToleranceDiameterPx(p.tolerance)}
            color={p.color}
            disabled={disabled}
          />
        ))}
      </div>
    </ResponsiveImage>
  )
}

export default PinImage
