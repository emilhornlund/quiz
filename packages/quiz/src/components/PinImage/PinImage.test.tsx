import { QuestionPinTolerance } from '@quiz/common'
import { render } from '@testing-library/react'
import React from 'react'
import { describe, expect, it } from 'vitest'

import PinImage from './PinImage'
import { PinColor } from './types.ts'

describe('PinImage', () => {
  it('should render PinImage with default props', async () => {
    const { container } = render(<PinImage />)

    expect(container).toMatchSnapshot()
  })

  it('should render PinImage with all props', async () => {
    const { container } = render(
      <PinImage
        imageURL="https://example.com/image.png"
        value={{ x: 0.1, y: 0.9, tolerance: QuestionPinTolerance.Medium }}
        values={[
          { x: 0.5, y: 0.45, color: PinColor.Green },
          { x: 0.3, y: 0.25, color: PinColor.Red },
          { x: 0.85, y: 0.75, color: PinColor.Red },
        ]}
        onChange={() => undefined}
        onValid={() => undefined}
      />,
    )

    expect(container).toMatchSnapshot()
  })

  it('should render PinImage with multiple pins', async () => {
    const { container } = render(
      <PinImage
        imageURL="https://example.com/image.png"
        values={[
          { x: 0.5, y: 0.45, color: PinColor.Green },
          { x: 0.3, y: 0.25, color: PinColor.Red },
          { x: 0.85, y: 0.75, color: PinColor.Red },
        ]}
        onChange={() => undefined}
        onValid={() => undefined}
      />,
    )

    expect(container).toMatchSnapshot()
  })
})
