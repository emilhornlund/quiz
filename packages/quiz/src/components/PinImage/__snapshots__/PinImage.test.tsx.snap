// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PinImage > should render PinImage with all props 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="centerOverlay"
    >
      <div
        class="loadingSpinnerContainer"
        data-testid="loading-spinner"
      >
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`PinImage > should render PinImage with default props 1`] = `
<div>
  <div
    class="container"
  />
</div>
`;

exports[`PinImage > should render PinImage with multiple pins 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="centerOverlay"
    >
      <div
        class="loadingSpinnerContainer"
        data-testid="loading-spinner"
      >
        <div />
        <div />
        <div />
      </div>
    </div>
  </div>
</div>
`;
